File: ts/src/redeem.ts
/**
 * Redeem (Miden-side of Flow C) — high-level helpers.
 *
 * Mirror of `darwin_sdk::redeem`. RedeemRequest carries the burn
 * amount and the recipient Miden account that will receive the
 * underlyings.
 */

import type { BasketHandle } from "./baskets";
import { SdkError } from "./deposit";

export interface RedeemRequest {
  readonly basket: BasketHandle;
  /** Basket-token base units to burn. */
  readonly burnAmount: bigint;
  /** Miden account that will receive the underlyings. */
  readonly recipientAccountId: string;
  readonly expiryBlock: bigint;
}

/**
 * Same checks as the Rust SDK: handle resolved, burn amount > 0,
 * recipient looks like a 0x-prefixed account ID.
 */
export function validateRedeemRequest(req: RedeemRequest): void {
  if (req.basket.protocolAccountId === null || req.basket.basketTokenFaucetId === null) {
    throw new SdkError(
      "not-connected",
      `basket ${req.basket.symbol} is not fully resolved on-chain yet`,
    );
  }
  if (req.burnAmount <= 0n) {
    throw new SdkError(
      "invalid-input",
      `burn amount must be > 0 (got ${req.burnAmount})`,
    );
  }
  if (!req.recipientAccountId.startsWith("0x")) {
    throw new SdkError(
      "invalid-input",
      `recipient account id must be 0x-prefixed (got ${req.recipientAccountId})`,
    );
  }
}

File: rust/src/redeem.rs
//! Redeem (Miden-side of Flow C) — high-level helpers.

use crate::BasketHandle;

/// A user's intent to redeem from a basket.
#[derive(Debug, Clone)]
pub struct RedeemRequest {
    pub basket: BasketHandle,
    /// Amount of basket token to burn, in basket-token base units.
    pub burn_amount: u64,
    /// Miden account id that receives the underlyings. For ETH-user
    /// flows (M2), this is the Miden Guardian relay wallet that then
    /// chains B2AGG notes to bridge each underlying to L1.
    pub recipient_account_id: u64,
    pub expiry_block: u64,
}
